dot/state/hashtoblock.go

```diff
@@ -0,0 +1,45 @@
+// Copyright 2022 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package state
+
+import (
+	"sync"
+
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/ChainSafe/gossamer/lib/common"
+)
+
+type hashToBlock struct {
+	mutex   sync.RWMutex
+	mapping map[common.Hash]*types.Block
+}
+
+func newHashToBlock() *hashToBlock {
+	return &hashToBlock{
+		mapping: make(map[common.Hash]*types.Block),
+	}
+}
+
+func (h *hashToBlock) get(hash common.Hash) (block *types.Block) {
+	h.mutex.RLock()
```

Comment:
No we need to lock the actual `map` so that it's not read and written at the same time (even at different keys) or it would panic.

Essentially this sums it up:

> Before Golang 1.6, concurrent read is OK, concurrent write is not OK, but write and concurrent read is OK. Since Golang 1.6, map cannot be read when it's being written. 

From https://stackoverflow.com/a/38140573